### ISSUE TITLE
pam_tacplus not working - logs error "PAM unable to dlopen & PAM adding faulty module - cannot open shared object file" tacc does not build and "-ldcrypto -ldtacc" not in "dependency_libs=" in files "libtac.la & pam_tacplus.la"  on clean OS installs: Debian 5-8

### ISSUE THREAD

**Issue by alextomko** [jeroennijhof/pam_tacplus#101]
I have configured /etc/pam.d/sshd with this at the top:

`auth include tacacs`

I have configured /etc/pam.d/tacacs with this:

```
#%PAM-1.0
    auth       sufficient   /usr/local/lib/security/pam_tacplus.so debug server=xxx secret=xxx server=xxx secret=xxx
    account    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=xxx secret=xxx server=xxx secret=xxx service=shell protocol=ssh
    session    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=xxx secret=xxx server=xxxx secret=xxx service=shell protocol=ssh
```

It never even tries to talk to the server - I run a tcpdump. 

This is what I see in `/var/log/auth.log` - it fails to try to even talk to the tacacs server and the acs server never receives tcp traffic from linux. Notably I have another server with exactly the same identical working configuration - it happens to be that no other fresh installed debian 8.8 server works - only got lucky with server #1.

```
sshd[6459]: PAM unable to dlopen(/usr/local/lib/security/pam_tacplus.so): libtac.so.2: cannot open shared object file: No such file or directory
sshd[6459]: PAM adding faulty module: /usr/local/lib/security/pam_tacplus.so
sshd[6459]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=  user=
```

**Comment by pprindeville** [comment]
Okay, so where _did_ `pam_tacplus.so` get installed?

**Comment by cbpsnetworking** [comment]
It is in `/usr/local/lib/security/pam_tacplus.so`.

**Comment by cbpsnetworking** [comment]
I tried 5 more clean install updated debian 8.8 computers and they all get the same error. I am going to assume some update or other has broken this package for Debian. 

Dec 26 14:27:18 localhost sshd[5129]: PAM unable to dlopen(/usr/local/lib/security/pam_tacplus.so): libtac.so.2: cannot open shared object file: No such file or directory
Dec 26 14:27:18 localhost sshd[5129]: PAM adding faulty module: /usr/local/lib/security/pam_tacplus.so

**Comment by pprindeville** [comment]
> I am going to assume some update or other has broken this package for Debian.

@cbpsnetworking I'd start with owner/group/permission on that file versus other PAM plugins, and then if that doesn't work try running `strace` on PAM.

I'd also look at what `pam_tacplus.so` exports and imports versus other modules.  There might be a dependency on a deprecated PAM export, or PAM might have a new requirement that an additional symbol gets exported by any plug-ins, etc.

**Comment by cbpsnetworking** [comment]
I checked all permissions/group ect - it is identical on the other 3 servers as it is on the broken server.

Working # 1:

@host1:/usr/local/lib$ ls -l /usr/local/lib
total 124
-rwxr-xr-x 1 root staff    948 Dec 21 09:10 libtac.la
lrwxrwxrwx 1 root staff     15 Dec 21 09:10 libtac.so -> libtac.so.2.0.0
lrwxrwxrwx 1 root staff     15 Dec 21 09:10 libtac.so.2 -> libtac.so.2.0.0
-rwxr-xr-x 1 root staff 104224 Dec 21 09:10 libtac.so.2.0.0
drwxr-sr-x 2 root staff   4096 Dec 21 09:10 pkgconfig
drwxrwsr-x 4 root staff   4096 Jul  1 09:51 python2.7
drwxrwsr-x 3 root staff   4096 Jul  1 09:54 python3.4
drwxr-sr-x 2 root staff   4096 Dec 21 09:10 security

Working #2:

@host2:~$ ls -l /usr/local/lib/security/
total 80
-rwxr-xr-x 1 root staff  1003 Dec 21 09:10 pam_tacplus.la
-rwxr-xr-x 1 root staff 75144 Dec 21 09:10 pam_tacplus.so
eraddatz@slcjump:~$ ls -l /usr/local/lib
total 124
-rwxr-xr-x 1 root staff    948 Dec 21 09:10 libtac.la
lrwxrwxrwx 1 root staff     15 Dec 21 09:10 libtac.so -> libtac.so.2.0.0
lrwxrwxrwx 1 root staff     15 Dec 21 09:10 libtac.so.2 -> libtac.so.2.0.0
-rwxr-xr-x 1 root staff 104224 Dec 21 09:10 libtac.so.2.0.0
drwxr-sr-x 2 root staff   4096 Dec 21 09:10 pkgconfig
drwxrwsr-x 4 root staff   4096 Jul  1 09:51 python2.7
drwxrwsr-x 3 root staff   4096 Jul  1 09:54 python3.4
drwxr-sr-x 2 root staff   4096 Dec 21 09:10 security


Non-working:

@non-host:/usr/local/lib/security$ ls -l /usr/local/lib/security/
total 80
-rwxr-xr-x 1 root staff   984 Dec 26 14:14 pam_tacplus.la
-rwxr-xr-x 1 root staff 76992 Dec 26 14:14 pam_tacplus.so
eraddatz@slcwks02:/usr/local/lib/security$ ls -l /usr/local/lib
total 144
-rwxr-xr-x 1 root staff    929 Dec 26 14:14 libtac.la
lrwxrwxrwx 1 root staff     15 Dec 26 14:14 libtac.so -> libtac.so.2.0.0
lrwxrwxrwx 1 root staff     15 Dec 26 14:14 libtac.so.2 -> libtac.so.2.0.0
-rwxr-xr-x 1 root staff 125832 Dec 26 14:14 libtac.so.2.0.0
drwxr-sr-x 2 root staff   4096 Dec 26 14:14 pkgconfig
drwxrwsr-x 4 root staff   4096 Jul 28 14:54 python2.7
drwxrwsr-x 3 root staff   4096 Jul 28 14:52 python3.5
drwxr-sr-x 2 root staff   4096 Dec 26 14:14 security

I do see file sizes a bit different on non-working to working. I have installed with this:

$ git clone https://github.com/jeroennijhof/pam_tacplus
$ sudo apt-get update
$ sudo apt-get install dh-autoreconf
$ sudo apt-get install libpam-dev
$ ./auto.sh
$ ./configure && make && sudo make install

The installation goes well each time. I also tried your fork once and it had same result.

**Comment by cbpsnetworking** [comment]
I tried another clean install of debian 8.8 today and same result. I will try debian 8.9 and 9.0 also now.

**Comment by cbpsnetworking** [comment]
I tried Debian 8, Debian 7, Debian 6, Debian5, CentOS7, CentOS6 all with clean installations and no updates this morning - all have exact same issue. I think I found what the issue is. There is some package on the working server that I have - which I have no idea what it is -  that builds the pam_tacplus directory properly with the -ltac -lcrypto. 

The `libtac.la` and `pam_tacplus.la` files in the build directory and the `/usr/local/lib/` directory are missing the `-ltac and `lcrypto`. It also does not build the `tacc` program and If I try to manually run it it says that I am missing `lcrypto`. 

So If I can get these libraries to build properly then it should fix the issue but how I am I supposed to find this if it does not just build with this package? I spent considerable time trying to figure this out but it appears default installations of all Debian OS and CentOS do not work out of the box? Help Please?

tacc: line 202: cd: /netops/docs/pam_tacplus: No such file or directory
/usr/bin/ld: cannot find -lcrypto

**Comment by pprindeville** [comment]
Sounds like you're missing `openssl`.  Go ahead and install that.  Then rebuild.

When you run:

```
% ldd /usr/local/lib/pam_tacplus.so
```

on the server which works, what do you get?

**Comment by cbpsnetworking** [comment]
OpenSSL comes pre-installed with all Debian OS though already.

working server:

~~~
$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 (0x00007ffc6b5f7000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f0bbfa08000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f0bbf60c000)
	libtac.so.2 => /usr/local/lib/libtac.so.2 (0x00007f0bbf406000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007f0bbf1f6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0bbee4b000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0bbec47000)
	libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007f0bbea21000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0bbfe12000)

~~~

non-working debian 8.8.0:

~~~
$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 (0x00007ffec39f1000)
	libtac.so.2 => not found
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fde5ec99000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007fde5ea89000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fde5e6de000)
	libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007fde5e4b8000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fde5e2b4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fde5f0a3000)
~~~

It appears I have openssl on all:

debian 8.8.0 - non working.

~~~
user@debian-8:~$ apt-cache show openssl
Package: openssl
Version: 1.0.1t-1+deb8u7
~~~

debian 8.8.0 - working:

~~~
$ apt-cache show openssl
Package: openssl
Version: 1.0.1t-1+deb8u7
Installed-Size: 1093
~~~

**Comment by pprindeville** [comment]
> OpenSSL comes pre-installed with all Debian OS though already.

@cbpsnetworking Does it?  Not seeing it on the list for the non-working Debian 8.8.0.

But you're obfuscating two different issues.

There are run-time dependencies, where `openssl` needs to be on the target machine _to run_.  There's the build-time dependency where `libssl-dev` needs to be on the build host _to build PAM_.

**Comment by pprindeville** [comment]
> The installation goes well each time. I also tried your fork once and it had same result.

You left out:

```
% sudo apt-get install libssl-dev
```

before doing the build.

**Comment by cbpsnetworking** [comment]
Ok - I did install `libssl-dev` just now - it is getting closer. It is now only missing `libtac.so.2` when I run ldd. I am not sure why it is missing it - the file is in `/usr/local/lib/libtac.so.2`.

~~~
user@debian-8:/netops/pam_tacplus$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 (0x00007ffd9b391000)
	libtac.so.2 => not found
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f34f1d43000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f34f1947000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007f34f1737000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f34f138c000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f34f1188000)
	libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007f34f0f62000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f34f214d000)
~~~

**Comment by pprindeville** [comment]
Did you **rebuild** and **reinstall**?

**Comment by pprindeville** [comment]
```
$ tacc
tacc: error while loading shared libraries: libtac.so.2: cannot open shared object file: No such file or directory
```

Okay, dumb question...  What do you get when you run:

```
% which tacc
```

and if you run:

```
% ./tacc
```

Do you get the same error?

**Comment by cbpsnetworking** [comment]
Yes I did - I deleted the build dir `pam_tacplus`. Git clone again of `https://github.com/jeroennijhof/pam_tacplus/` and also deleted /usr/local/lib/*. Then I did the above `./auto.sh ./configure && make && sudo make install`. I also rebooted and tried it a second time.

**Comment by cbpsnetworking** [comment]
*Could be getting confusing - just a recap of my Exact Steps - Install Clean Install Debian versions 7-8-9 with openssh selected:*
~~~
Add host IP to an tacacs+ server for later.
~~~

*Install - dh-autoreconf, libpam-dev, libssl-dev, & check that openssl is installed:*
~~~
$ sudo apt-get install dh-autoreconf
$ sudo apt-get install libpam-dev
$ sudo apt-get install libssl-dev
$ sudo apt-get install openssl
$ git clone https://github.com/jeroennijhof/pam_tacplus/
$ cd pam_tacplus
$ ./auto.sh
$ ./configure && make && sudo make install
~~~

*add `auth include tacacs` at top of `/etc/pam.d/sshd`:*
*add file `/etc/pam.d/tacacs` with:*
~~~
#%PAM-1.0
    auth       sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret
    account    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
    session    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
~~~

*Check libraries: `libtac.so.2` missing:*
~~~
$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 (0x00007fff401e7000)
	libtac.so.2 => not found
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f5d587c5000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f5d583c9000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007f5d581b9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5d57e0e000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5d57c0a000)
	libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007f5d579e4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5d58bcf000)
~~~ 
*Check if tacc runs:*
~~~
$ tacc
tacc: error while loading shared libraries: libtac.so.2: cannot open shared object file: No such file or directory
~~~

*Try an SSH to our host - with our username that has empty password:*
~~~
$ sudo cat /var/log/auth.log
Dec 28 11:52:51 debian-8 sshd[446]: PAM unable to dlopen(/usr/local/lib/security/pam_tacplus.so): /usr/local/lib/security/pam_tacplus.so: cannot open shared object file: No such file or directory
Dec 28 12:44:15 debian-8 sshd[9870]: PAM unable to dlopen(/usr/local/lib/security/pam_tacplus.so): libtac.so.2: cannot open shared object file: No such file or directory
~~~

**Comment by cbpsnetworking** [comment]
I am not sure how I found this or what it means but by doing this exact step - it started working:

Notice that after I run `ldconfig -v 2>/dev/null | grep -v ^$'\t'`` - I run ldd on `pam_tacplus.so` and it shows the `libtac.so.2` library. 

I then try tacacs user-id from SSH and it works. I also rebooted both machines and the configuration stayed.

~~~
$ sudo -s # get into root 
# ldconfig -v 2>/dev/null | grep -v ^$'\t'` 
# ldd /usr/local/bin/security/pam_tacplus.so
        linux-vdso.so.1 =>  (0x00007ffcfa7d1000)
	libtac.so.2 => /usr/local/lib/libtac.so.2 (0x00007fb53ef62000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fb53ed5f000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007fb53e965000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007fb53e757000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb53e3ca000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fb53e1c6000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb53dfaf000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb53f36f000)
# exit

$ ldd /usr/local/lib/security/pam_tacplus.so        
linux-vdso.so.1 =>  (0x00007ffcfa7d1000)
	libtac.so.2 => /usr/local/lib/libtac.so.2 (0x00007fb53ef62000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fb53ed5f000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007fb53e965000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007fb53e757000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb53e3ca000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fb53e1c6000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb53dfaf000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb53f36f000)
~~~

SSH works now on Debian 8/7 and before:

~~~
Dec 28 15:01:07 debian-8 sshd[446]: pam_sm_setcred: called (pam_tacplus v1.3.8)
Dec 28 15:01:07 debian-8 sshd[446]: pam_unix(sshd:session): session opened for user netadmin by (uid=0)
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: 1 servers defined
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: server[0] { addr=172.16.1.115:49, key='' }
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: tac_service=''
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: tac_protocol=''
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: tac_prompt=''
Dec 28 15:01:07 debian-8 PAM-tacplus[448]: tac_login=''
Dec 28 15:01:07 debian-8 sshd[448]: pam_sm_setcred: called (pam_tacplus v1.3.8)
Dec 28 15:02:09 debian-8 sudo: netadmin : TTY=pts/0 ; PWD=/home/netadmin ; USER=root ; COMMAND=/bin/cat /var/log/auth.log
Dec 28 15:02:09 debian-8 sudo: pam_unix(sudo:session): session opened for user root by netadmin(uid=0)
netadmin@debian-8:~$ 
~~~

~~~
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: 2 servers defined
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: server[0] { addr=172.16.1.115:49, key='' }
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: server[1] { addr=172.17.1.115:49, key='' }
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: tac_service=''
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: tac_protocol=''
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: tac_prompt=''
Dec 28 15:02:38 debian-7 PAM-tacplus[1973]: tac_login=''
Dec 28 15:02:38 debian-7 sshd[1973]: pam_sm_setcred: called (pam_tacplus v1.3.8)
Dec 28 15:02:42 debian-7 sudo: netadmin : TTY=pts/0 ; PWD=/home/netadmin ; USER=root ; COMMAND=/bin/cat /var/log/auth.log
Dec 28 15:02:42 debian-7 sudo: pam_unix(sudo:session): session opened for user root by netadmin(uid=0)
~~~

**Comment by cbpsnetworking** [comment]
*All Steps related to how I got Clean Installs of Debian6-8 working*

*Install Clean Install Debian 6-7-8 with openssh selected:*
~~~
Make my user a sudo user with nopasswd:
Add host to an tacacs+ server for later.
~~~

*Install - sudo, dh-autoreconf, libpam-dev, libssl-dev, check that openssl is installed, install git&git-core:*
~~~
# apt-get install sudo
$ sudo apt-get install dh-autoreconf
$ sudo apt-get install libpam-dev
$ sudo apt-get install libssl-dev
$ sudo apt-get install openssl
$ sudo apt-get install git && sudo apt-get install git-core
$ git clone https://github.com/jeroennijhof/pam_tacplus/
$ cd pam_tacplus
$ ./auto.sh
$ ./configure && make && sudo make install
# You have to run `ldconfig` string or `libtac.so.2` library path is missing below and you get a `dlopen` error in `/var/log/auth.log` saying it cant find `libtac.so.2`.
$ sudo -s
# ldconfig -v 2>/dev/null | grep -v ^$'\t'
*Now all libraries work `libtac.so.2` showing now:*
$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 (0x00007fff6a76c000)
	libtac.so.2 => /usr/local/lib/libtac.so.2 (0x00007f36e8904000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f36e8701000)
	libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f36e8305000)
	libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007f36e80f5000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f36e7d4a000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f36e7b46000)
	libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007f36e7920000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f36e8d11000)
*SSH with password-less user works now:*
~~~

*add `auth include tacacs` at top of `/etc/pam.d/sshd`:*
*add file `/etc/pam.d/tacacs` with:*
~~~
#%PAM-1.0
    auth       sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret
    account    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
    session    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
~~~

*Check if tacc runs:*
~~~
$ tacc
tacc -- simple TACACS+ client and login, version 1.9.0
!Omitted
~~~

*Try an SSH to our host - with our username that has empty password:*
~~~
ssh userid@host
password:
host$
~~~

**Comment by cbpsnetworking** [comment]
You will notice that I took Debian 9 out of this. It no longer even builds properly after installing `libssl-dev`. The `./configure` step works but `make` has an `openssl` error. I am not concerned about fixing Debian version 9 at this time.

**Comment by pprindeville** [comment]
It's normal to have to run `ldconfig` to update the loader cache.

```
$ rpm -q --scripts openssl-libs
postinstall program: /sbin/ldconfig
postuninstall program: /sbin/ldconfig
$ 
```

**Comment by pprindeville** [comment]
If you're going to mention the Debian 9 issue you could at least attach logs in case it rings bells with anyone else or if it's low-hanging fruit.

Note: after installing shared libraries anywhere in the standard `$LD_LIBRARY_PATH`, it's a given that you have to run `sudo /sbin/ldconfig`.  That's understood.

Also note that (1) you can install multiple packages on a single `apt-get install` line, and (2) installing `git` installs `git-core` which is a dependency, so you never need to install both.  You also said that `openssl` is pre-installed on a pristine system, so you shouldn't have to explicitly (re)install that either.

**Comment by cbpsnetworking** [comment]
pprindeville - thanks for your help. It was not understood to me and I could not find any information on how to get it running. My goal here was to do clean installation which I did just yesterday/day before and gather the steps to get the pam_tacplus working and thanks to your help I now am able to do that for both Debian and CentOS. Appreciate it very much. I will post the Debian 9 build issue.

**Comment by cbpsnetworking** [comment]
On Debian 9.3.0 OS - clean installation with the required packages ` dh-autoreconf libpam-dev libssl-dev git`

It built fine before installing the `libssl-dev`, once you install this package you get an error after the `./configure` step while it is doing `make`.

It end ends with `Error 2`. I attached full log also.

~~~
libtac/lib/magic.c: In function ‘magic’:
libtac/lib/magic.c:66:5: error: ‘RAND_pseudo_bytes’ is deprecated [-Werror=deprecated-declarations]
     RAND_pseudo_bytes((unsigned char *)&num, sizeof(num));
     ^~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/e_os2.h:13:0,
                 from /usr/include/openssl/ossl_typ.h:19,
                 from /usr/include/openssl/rand.h:14,
                 from libtac/lib/magic.c:56:
/usr/include/openssl/rand.h:47:1: note: declared here
 DEPRECATEDIN_1_1_0(int RAND_pseudo_bytes(unsigned char *buf, int num))
 ^
cc1: all warnings being treated as errors
Makefile:825: recipe for target 'libtac/lib/libtac_la-magic.lo' failed
make[1]: *** [libtac/lib/libtac_la-magic.lo] Error 1
make[1]: Leaving directory '/netops/pam_tacplus'
Makefile:458: recipe for target 'all' failed
make: *** [all] Error 2

~~~


[pam_tacplus-build.txt](https://github.com/jeroennijhof/pam_tacplus/files/1593354/pam_tacplus-build.txt)

**Comment by pprindeville** [comment]
@cbpsnetworking This has been fixed in PR #105.  Please retry.

**Comment by alextomko** [comment]
Let's include Debian 9-11 now on this issue.

~~~
Jun  3 21:29:10 azwcnojump sshd[34491]: PAM unable to dlopen(/usr/local/lib/security/pam_tacplus.so): libtac.so.5: cannot open shared object file: No such file or directory
~~~

Fresh install Debian 11
